File: iOS 8 by Tutorials/Objective-C/Chapter 20 - Intermediate Photos/Stitch/Stitch/Helpers/StitchHelper.m
//
//  StitchCreator.m
//  Stitch
//
//  Created by Jack Wu on 2014-06-29.
//
//

@import CoreGraphics;
@import UIKit;
#import "StitchHelper.h"

#define StitchWidth 900
#define MaxPhotosPerStitch 6

@implementation StitchHelper

+ (void)createNewStitchWith:(NSArray *)assets inCollection:(PHAssetCollection *)collection {
  UIImage * stitchImage = [StitchHelper createStitchImageWithAssets:assets];
  
  __block PHObjectPlaceholder * stitchPlaceholder;
  [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
    PHAssetChangeRequest * assetChangeRequest = [PHAssetChangeRequest creationRequestForAssetFromImage:stitchImage];
    
    stitchPlaceholder = assetChangeRequest.placeholderForCreatedAsset;
    
    PHAssetCollectionChangeRequest *assetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:collection assets:nil];
    [assetCollectionChangeRequest addAssets:@[stitchPlaceholder]];
    
  } completionHandler:^(BOOL success, NSError *error) {
    // Fetch the new asset and add the adjustment data to it
    PHFetchResult * fetch = [PHAsset fetchAssetsWithLocalIdentifiers:@[stitchPlaceholder.localIdentifier] options:nil];
    PHAsset * newStitch = fetch[0];
    [self editStitchContentWith:newStitch image:stitchImage assets:assets];
  }];
}

+ (void)editStitchContentWith:(PHAsset *)stitch image:(UIImage *)image assets:(NSArray *)assets {
  NSData * stitchJPEG = UIImageJPEGRepresentation(image, 0.9);
  NSArray * assetsIDs = [assets valueForKey:@"localIdentifier"];
  NSData * assetsData = [NSKeyedArchiver archivedDataWithRootObject:assetsIDs];
  
  [stitch requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput *contentEditingInput, NSDictionary *info) {
    PHAdjustmentData * adjustmentData = [[PHAdjustmentData alloc] initWithFormatIdentifier:StitchAdjustmentFormatIdentifier formatVersion:@"1.0" data:assetsData];
    
    PHContentEditingOutput * editingOutput = [[PHContentEditingOutput alloc] initWithContentEditingInput:contentEditingInput];
    editingOutput.adjustmentData = adjustmentData;
    [stitchJPEG writeToURL:editingOutput.renderedContentURL atomically:YES];
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
      PHAssetChangeRequest *request = [PHAssetChangeRequest changeRequestForAsset:stitch];
      request.contentEditingOutput = editingOutput;
    } completionHandler:^(BOOL success, NSError *error) {
      if (!success) {
        NSLog(@"Error: %@", error);
      }
    }];
  }];
}

+ (void)loadAssetsInStitch:(PHAsset *)stitch completion:(void(^)(NSMutableArray *assets))completion {
  PHContentEditingInputRequestOptions *options = [[PHContentEditingInputRequestOptions alloc] init];
  [options setCanHandleAdjustmentData:^BOOL(PHAdjustmentData *adjustmentData) {
    return [adjustmentData.formatIdentifier isEqualToString:StitchAdjustmentFormatIdentifier] && [adjustmentData.formatVersion isEqualToString:@"1.0"];
  }];
  
  [stitch requestContentEditingInputWithOptions:options completionHandler:^(PHContentEditingInput *contentEditingInput, NSDictionary *info) {
    if (contentEditingInput.adjustmentData) {
      NSArray * stitchAssetIDs = [NSKeyedUnarchiver unarchiveObjectWithData:contentEditingInput.adjustmentData.data];
      PHFetchResult * stitchAssetsFetchResult = [PHAsset fetchAssetsWithLocalIdentifiers:stitchAssetIDs options:nil];
      
      NSMutableArray *stitchAssets = [[NSMutableArray alloc] initWithCapacity:stitchAssetsFetchResult.count];
      [stitchAssetsFetchResult enumerateObjectsUsingBlock:^(PHAsset * obj, NSUInteger idx, BOOL *stop) {
        [stitchAssets addObject:obj];
        completion(stitchAssets);
      }];
    }
    else {
      completion(nil);
    }
  }];
}

+ (UIImage *)createStitchImageWithAssets:(NSArray *)assets {
  NSInteger assetCount = assets.count;
  
  if (!assetCount) {
    return nil;
  }
  
  if (assetCount > MaxPhotosPerStitch) {
    assetCount = MaxPhotosPerStitch;
  }
  
  NSArray * placementRects = [self placementRectsForAssetCount:assetCount];
  
  CGFloat deviceScale = [UIScreen mainScreen].scale;
  UIGraphicsBeginImageContextWithOptions(CGSizeMake(StitchWidth, StitchWidth), YES, deviceScale);
  
  PHImageRequestOptions * options = [[PHImageRequestOptions alloc] init];
  options.synchronous = YES;
  options.resizeMode = PHImageRequestOptionsResizeModeExact;
  options.deliveryMode = PHImageRequestOptionsDeliveryModeHighQualityFormat;
  
  for (int i = 0; i < assetCount; i++) {
    @autoreleasepool {
      PHAsset * asset = assets[i];
      CGRect rect = [placementRects[i] CGRectValue];
      CGSize targetSize = CGSizeMake(CGRectGetWidth(rect)*deviceScale, CGRectGetHeight(rect)*deviceScale);
      [[PHImageManager defaultManager] requestImageForAsset:asset targetSize:targetSize contentMode:PHImageContentModeAspectFill options:options resultHandler:^(UIImage *result, NSDictionary *info) {
        if (!CGSizeEqualToSize(result.size, targetSize)) {
          result = [self cropImage:result toSize:targetSize];
        }
        [result drawInRect:rect];
      }];
    }
  }
  
  UIImage * result = UIGraphicsGetImageFromCurrentImageContext();
  UIGraphicsEndImageContext();
  
  return result;
}

+ (NSArray *)placementRectsForAssetCount:(NSInteger)count {
  NSMutableArray * rects = [[NSMutableArray alloc] initWithCapacity:count];
  
  NSInteger evenCount;
  NSInteger oddCount;
  if (count % 2 == 0) {
    evenCount = count;
    oddCount = 0;
  }
  else {
    oddCount = 1;
    evenCount = count - oddCount;
  }
  
  CGFloat rectHeight = StitchWidth / (evenCount / 2 + oddCount);
  CGFloat evenWidth = StitchWidth / 2;
  CGFloat oddWidth = StitchWidth;
  
  for (int i = 0; i < evenCount; i++) {
    CGRect rect = CGRectMake(i%2 * evenWidth, i/2 * rectHeight, evenWidth, rectHeight);
    [rects addObject:[NSValue valueWithCGRect:rect]];
  }
  
  if (oddCount) {
    CGRect rect = CGRectMake(0, evenCount/2 * rectHeight, oddWidth, rectHeight);
    [rects addObject:[NSValue valueWithCGRect:rect]];
  }
  
  return rects;
}

+ (UIImage *)cropImage:(UIImage*)image toSize:(CGSize)size {
  CGFloat ratio = MIN(image.size.width / size.width, image.size.height / size.height);

  CGSize newSize = {image.size.width / ratio, image.size.height / ratio};
  CGPoint offset = {0.5 * (size.width - newSize.width), 0.5 * (size.height - newSize.height)};
  CGRect rect = {offset, newSize};
  
  UIGraphicsBeginImageContext(size);
  [image drawInRect:rect];
  UIImage * output = UIGraphicsGetImageFromCurrentImageContext();
  UIGraphicsEndImageContext();
  
  return output;
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 20 - Intermediate Photos/Stitch-Starter/Stitch/StitchHelper.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/


import UIKit
import Photos
import CoreGraphics

let StitchWidth = 900
let MaxPhotosPerStitch = 6

let StitchAdjustmentFormatIdentifier = "RW.stitch.adjustmentFormatID"

class StitchHelper: NSObject {
  
  // MARK: Stitch Creation
  class func createNewStitchWith(assets: [PHAsset], inCollection collection: PHAssetCollection) {
    
  }
  
  // MARK: Stitch Content
  class func editStitchContentWith(stitch: PHAsset, image: UIImage, assets: [PHAsset]) {
    
  }
  
  class func loadAssetsInStitch(stitch: PHAsset, completion: [PHAsset] -> ()) {
    
  }
  
  // MARK: Stitch Image Creation
  class func createStitchImageWithAssets(assets: [PHAsset]) -> UIImage {
    var assetCount = assets.count
    
    // Cap to 6 max photos
    if (assetCount > MaxPhotosPerStitch) {
      assetCount = MaxPhotosPerStitch
    }
    
    // Calculate placement rects
    let placementRects = placementRectsForAssetCount(assetCount)
    
    // Create context to draw images
    let deviceScale = UIScreen.mainScreen().scale
    UIGraphicsBeginImageContextWithOptions(CGSize(width: StitchWidth, height: StitchWidth), true, deviceScale)
    
    let options = PHImageRequestOptions()
    options.synchronous = true
    options.resizeMode = .Exact
    options.deliveryMode = .HighQualityFormat
    
    // Draw each image into their rect
    for (i: Int, asset: PHAsset) in enumerate(assets) {
      if (i >= assetCount) {
        break
      }
      let rect = placementRects[i]
      let targetSize = CGSize(width: CGRectGetWidth(rect)*deviceScale, height: CGRectGetHeight(rect)*deviceScale)
      PHImageManager.defaultManager().requestImageForAsset(asset, targetSize:targetSize, contentMode:.AspectFill, options:options) { result, _ in
        if result.size != targetSize {
          let croppedResult = self.cropImageToCenterSquare(result, size:targetSize)
          croppedResult.drawInRect(rect)
        } else {
          result.drawInRect(rect)
        }
      }
    }
    
    // Grab results
    let result = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    
    return result
  }
  
  private class func placementRectsForAssetCount(count: Int) -> [CGRect] {
    var rects: [CGRect] = []
    
    var evenCount: Int
    var oddCount: Int
    if count % 2 == 0 {
      evenCount = count
      oddCount = 0
    } else {
      oddCount = 1
      evenCount = count - oddCount
    }
    
    let rectHeight = StitchWidth / (evenCount / 2 + oddCount)
    let evenWidth = StitchWidth / 2
    let oddWidth = StitchWidth
    
    for i in 0..<evenCount {
      let rect = CGRect(x: i%2 * evenWidth, y: i/2 * rectHeight, width: evenWidth, height: rectHeight)
      rects.append(rect)
    }
    
    if oddCount > 0 {
      let rect = CGRect(x: 0, y: evenCount/2 * rectHeight, width: oddWidth, height: rectHeight)
      rects.append(rect)
    }
    
    return rects
  }
  
  // Helper to crop Image if it wasn't properly cropped
  private class func cropImageToCenterSquare(image: UIImage, size: CGSize) -> UIImage {
    let ratio = min(image.size.width / size.width, image.size.height / size.height)
    
    let newSize = CGSize(width: image.size.width / ratio, height: image.size.height / ratio)
    let offset = CGPoint(x: 0.5 * (size.width - newSize.width), y: 0.5 * (size.height - newSize.height))
    let rect = CGRect(origin: offset, size: newSize)
    
    UIGraphicsBeginImageContext(size)
    image.drawInRect(rect)
    let output = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    
    return output
  }
  
}
